File: app/src/main/java/com/pr0gramm/app/ui/views/InfoLineView.java
package com.pr0gramm.app.ui.views;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Ordering;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.api.pr0gramm.response.Tag;
import com.pr0gramm.app.feed.FeedItem;
import com.pr0gramm.app.feed.Vote;
import com.pr0gramm.app.ui.ConservativeLinearLayoutManager;
import com.pr0gramm.app.ui.MergeRecyclerAdapter;
import com.pr0gramm.app.ui.SingleViewAdapter;
import com.pr0gramm.app.ui.TagCloudLayoutManager;

import org.joda.time.Duration;
import org.joda.time.Instant;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import butterknife.ButterKnife;
import rx.Observable;
import rx.functions.Actions;

import static com.pr0gramm.app.Dagger.appComponent;
import static com.pr0gramm.app.util.AndroidUtility.checkMainThread;
import static com.pr0gramm.app.util.AndroidUtility.findView;
import static java.lang.Math.min;
import static net.danlew.android.joda.DateUtils.getRelativeTimeSpanString;

/**
 */
public class InfoLineView extends LinearLayout {
    private final VoteView voteView;
    private final TextView ratingView;
    private final TextView dateView;
    private final UsernameView usernameView;
    private final RecyclerView tagsView;
    private final Pr0grammIconView voteFavoriteView;

    @Nullable
    private final Settings settings;

    private final View ratingUnknownView;
    private final boolean admin;
    private OnDetailClickedListener onDetailClickedListener;
    private VoteView.OnVoteListener onVoteListener;

    private FeedItem feedItem;
    private TagVoteListener tagVoteListener;
    private boolean isSelfPost;
    private Runnable onAddTagClickedListener;
    private TagsAdapter tagsAdapter;

    public InfoLineView(Context context) {
        this(context, null);
    }

    public InfoLineView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public InfoLineView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        settings = isInEditMode() ? null : Settings.of(context);
        admin = !isInEditMode() && appComponent(context).userService().userIsAdmin();

        setOrientation(VERTICAL);

        inflate(context, R.layout.post_info_line, this);

        // get the views from the hierarchy
        ratingView = findView(this, R.id.rating);
        ratingUnknownView = findView(this, R.id.rating_hidden);
        usernameView = findView(this, R.id.username);
        dateView = findView(this, R.id.date);
        voteView = findView(this, R.id.voting);
        voteFavoriteView = findView(this, R.id.action_favorite);

        tagsView = findView(this, R.id.tags);

        int tagGaps = getResources().getDimensionPixelSize(R.dimen.tag_gap_size);
        if (settings != null && settings.tagCloudView()) {
            tagsView.setItemAnimator(null);
            tagsView.setLayoutManager(new TagCloudLayoutManager(tagGaps, tagGaps, 3));
        } else {
            tagsView.setItemAnimator(null);

            // tagsView.setLayoutManager(new TagCloudLayoutManager(tagGaps, tagGaps, 1));
            tagsView.setLayoutManager(new ConservativeLinearLayoutManager(getContext(), HORIZONTAL, false));
        }

        voteView.setOnVoteListener(newVote -> {
            boolean changed = onVoteListener != null && onVoteListener.onVoteClicked(newVote);
            if (changed)
                updateViewState(newVote);

            return changed;
        });

        voteFavoriteView.setOnClickListener(v -> {
            Vote currentVote = voteView.getVote();
            if (currentVote == Vote.FAVORITE) {
                voteView.setVote(Vote.UP);
            } else {
                voteView.setVote(Vote.FAVORITE);
            }
        });

        setTags(Collections.<Tag, Vote>emptyMap());
    }

    /**
     * Displays the given {@link com.pr0gramm.app.feed.FeedItem} along with
     * the given vote.
     *
     * @param item The item to display
     * @param vote The vote that belongs to the given item.
     */
    public void setFeedItem(FeedItem item, boolean isSelfPost, Observable<Vote> vote) {
        this.feedItem = item;
        this.isSelfPost = isSelfPost;

        // update the views!
        usernameView.setUsername(item.user(), item.mark());
        dateView.setText(getRelativeTimeSpanString(getContext(), item.created()));
        updateViewState(Vote.NEUTRAL);

        usernameView.setOnClickListener(v -> {
            if (onDetailClickedListener != null) {
                String username = item.user();
                onDetailClickedListener.onUserClicked(username);
            }
        });

        vote.subscribe(v -> {
            checkMainThread();
            voteView.setVote(v, true);

            updateViewState(v);
        }, Actions.empty());
    }

    /**
     * Updates the rating using the currently set feed item and the given vote.
     *
     * @param vote The vote that is currently selected.
     */
    private void updateViewState(Vote vote) {
        if (feedItem == null)
            return;

        if (isOneHourOld() || isSelfPost || admin) {
            int rating = feedItem.up() - feedItem.down() + min(1, vote.getVoteValue());
            ratingView.setText(String.valueOf(rating));
            ratingView.setOnLongClickListener(v -> {
                Toast.makeText(getContext(),
                        String.format("%d up, %d down", feedItem.up(), feedItem.down()),
                        Toast.LENGTH_SHORT).show();

                return true;
            });

            ratingView.setVisibility(VISIBLE);
            ratingUnknownView.setVisibility(GONE);

        } else {
            ratingUnknownView.setVisibility(VISIBLE);
            ratingView.setVisibility(GONE);
            ratingView.setOnLongClickListener(null);
        }

        voteFavoriteView.setTextColor(vote == Vote.FAVORITE
                ? voteView.getMarkedColor()
                : voteView.getDefaultColor());
    }

    public VoteView getVoteView() {
        return voteView;
    }

    public void setTags(Map<Tag, Vote> tags) {
        List<Tag> sorted = Ordering.natural().reverse()
                .onResultOf(Tag::getConfidence)
                .sortedCopy(tags.keySet());

        View addTagView = LayoutInflater.from(getContext()).inflate(R.layout.tags_add, null);
        addTagView.setOnClickListener(v -> {
            if (onAddTagClickedListener != null) {
                onAddTagClickedListener.run();
            }
        });

        addTagView.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));

        int spec = MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED);
        addTagView.measure(spec, spec);

        int height = addTagView.getMeasuredHeight();
        tagsView.setMinimumHeight(height);

        tagsAdapter = new TagsAdapter(sorted, tags);

        MergeRecyclerAdapter adapter = new MergeRecyclerAdapter();
        adapter.addAdapter(SingleViewAdapter.ofView(addTagView));
        adapter.addAdapter(tagsAdapter);
        tagsView.setAdapter(adapter);
    }

    public void setOnDetailClickedListener(OnDetailClickedListener onDetailClickedListener) {
        this.onDetailClickedListener = onDetailClickedListener;
    }

    public void setOnVoteListener(VoteView.OnVoteListener onVoteListener) {
        this.onVoteListener = onVoteListener;
    }

    public void setTagVoteListener(TagVoteListener tagVoteListener) {
        this.tagVoteListener = tagVoteListener;
    }

    public void setOnAddTagClickedListener(Runnable onAddTagClickedListener) {
        this.onAddTagClickedListener = onAddTagClickedListener;
    }

    public boolean isOneHourOld() {
        Instant oneHourAgo = Instant.now().minus(Duration.standardHours(1));
        return feedItem.created().isBefore(oneHourAgo);
    }

    public void addVote(Tag tag, Vote vote) {
        tagsAdapter.updateVote(tag, vote);
    }

    private class TagsAdapter extends RecyclerView.Adapter<TagViewHolder> {
        private final List<Tag> tags;
        private final Map<Tag, Vote> votes;
        private final boolean alwaysVoteViews;
        private int selected = -1;

        private TagsAdapter(List<Tag> tags, Map<Tag, Vote> votes) {
            setHasStableIds(true);
            this.tags = ImmutableList.copyOf(tags);
            this.votes = new HashMap<>(votes);
            this.alwaysVoteViews = settings != null && !settings.hideTagVoteButtons();
        }

        @Override
        public TagViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            View view = inflater.inflate(R.layout.tag, parent, false);
            return new TagViewHolder(view);
        }

        @Override
        public void onBindViewHolder(TagViewHolder holder, int position) {
            Tag tag = tags.get(position);
            holder.tag.setText(tag.getTag());
            holder.tag.setOnClickListener(v -> {
                if (onDetailClickedListener != null)
                    onDetailClickedListener.onTagClicked(tag);
            });

            if (tagVoteListener != null && shouldShowVoteView(position)) {
                holder.vote.setVote(votes.get(tag), true);
                holder.vote.setVisibility(View.VISIBLE);

                if (!alwaysVoteViews) {
                    holder.tag.setOnLongClickListener(v -> {
                        updateSelection(-1);
                        return true;
                    });
                }

                holder.vote.setOnVoteListener(vote -> tagVoteListener.onVote(tag, vote));

            } else {
                holder.vote.setVisibility(View.GONE);
                holder.tag.setOnLongClickListener(v -> {
                    updateSelection(position);
                    return true;
                });
            }
        }

        private boolean shouldShowVoteView(int position) {
            return position == selected || alwaysVoteViews;
        }

        private void updateSelection(int position) {
            int previousSelected = selected;
            selected = position;

            notifyItemChanged(Math.max(0, previousSelected));
            notifyItemChanged(selected);
        }

        @Override
        public int getItemCount() {
            return tags.size();
        }

        @Override
        public long getItemId(int position) {
            return tags.get(position).getId();
        }

        public void updateVote(Tag tag, Vote vote) {
            votes.put(tag, vote);
            notifyDataSetChanged();
        }
    }

    private static class TagViewHolder extends RecyclerView.ViewHolder {
        final TextView tag;
        final VoteView vote;

        TagViewHolder(View itemView) {
            super(itemView);
            tag = ButterKnife.findById(itemView, R.id.tag_text);
            vote = ButterKnife.findById(itemView, R.id.tag_vote);
        }
    }

    public interface OnDetailClickedListener {
        /**
         * Called if the user clicked on a tag.
         *
         * @param tag The tag that was clicked.
         */
        void onTagClicked(Tag tag);

        /**
         * Called if a user clicks on a username
         *
         * @param username The username that was clicked.
         */
        void onUserClicked(String username);
    }

    public interface TagVoteListener {
        boolean onVote(Tag tag, Vote vote);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/InfoLineView.kt
package com.pr0gramm.app.ui.views

import android.annotation.SuppressLint
import android.content.Context
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.widget.PopupMenu
import androidx.appcompat.widget.menuPopupHelper
import androidx.core.view.isVisible
import com.google.android.material.snackbar.Snackbar
import com.pr0gramm.app.Duration
import com.pr0gramm.app.Instant
import com.pr0gramm.app.R
import com.pr0gramm.app.Settings
import com.pr0gramm.app.feed.FeedItem
import com.pr0gramm.app.orm.Vote
import com.pr0gramm.app.services.FollowState
import com.pr0gramm.app.services.SingleShotService
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.ui.DrawableCache
import com.pr0gramm.app.ui.base.whileIsAttachedScope
import com.pr0gramm.app.ui.base.withErrorDialog
import com.pr0gramm.app.ui.configureNewStyle
import com.pr0gramm.app.ui.showDialog
import com.pr0gramm.app.util.*
import com.pr0gramm.app.util.di.injector
import java.text.DateFormat
import kotlin.math.min


/**
 */
class InfoLineView(context: Context) : LinearLayout(context), InjectorViewMixin {
    init {
        inflate(context, R.layout.post_info_line, this)
    }

    private val singleShotService: SingleShotService by instance()

    private val captionView: TextView = find(R.id.date)
    private val usernameView: UsernameView = find(R.id.username)
    private val followStateView: ImageView = find(R.id.action_follow)
    private val collectionView: CollectView = find(R.id.collect)

    private val admin: Boolean = injector?.instance<UserService>()?.userIsAdmin == true

    private var feedItem: FeedItem? = null
    private var isSelfPost: Boolean = false

    private val voteController = VoteViewController(find(R.id.vote_up), find(R.id.vote_down))

    var onDetailClickedListener: PostActions? = null

    init {
        orientation = VERTICAL

        voteController.onVoteClicked = { newVote ->
            val changed = onDetailClickedListener?.votePostClicked(newVote) ?: false
            changed.also { updateViewState(newVote) }
        }

        followStateView.setOnClickListener {
            val iconColor = context.getStyledColor(android.R.attr.textColorSecondary)

            val popup = PopupMenu(context, followStateView)
            popup.menuPopupHelper.setForceShowIcon(true)
            popup.inflate(R.menu.menu_follow)
            popup.setOnMenuItemClickListener { followMenuClicked(it.itemId); true }

            popup.menu.findItem(R.id.action_follow_off)?.icon = DrawableCache
                    .get(R.drawable.ic_action_follow_off, iconColor)

            popup.menu.findItem(R.id.action_follow_normal)?.icon = DrawableCache
                    .get(R.drawable.ic_action_follow_normal, iconColor)

            popup.menu.findItem(R.id.action_follow_full)?.icon = DrawableCache
                    .get(R.drawable.ic_action_follow_full, iconColor)

            popup.show()
        }

        captionView.setOnLongClickListener { _ ->
            val item = this.feedItem
            if (item != null && scoreIsVisible()) {
                val date = item.created.toString(DateFormat.getDateTimeInstance())
                val text = "${item.up} Blussies, ${item.down} Minus\nErstellt am $date"

                Snackbar.make(this, text, Snackbar.LENGTH_SHORT)
                        .configureNewStyle()
                        .setAction(R.string.okay) {}
                        .show()
            }

            true
        }

        collectionView.setOnClickListener {
            collectClicked()
        }

        collectionView.setOnLongClickListener {
            onDetailClickedListener?.showCollectionsClicked()
            true
        }
    }

    private fun collectClicked() {
        if (!collectionView.isCollected && voteController.currentVote !== Vote.UP) {
            when {
                Settings.upvoteOnCollect -> {
                    triggerUpvoteOnCollect()
                }

                singleShotService.markAsDoneOnce("hint:upvoteOnCollect") -> {
                    askUpvoteOnCollect()
                }
            }
        }

        onDetailClickedListener?.collectClicked()
    }

    private fun askUpvoteOnCollect() {
        showDialog(context) {
            content(R.string.hint_upvote_on_collect)
            negative(android.R.string.no)
            positive(android.R.string.yes) {
                Settings.edit {
                    putBoolean("pref_upvote_on_collect", true)
                }

                // upvote this post too
                triggerUpvoteOnCollect()
            }
        }
    }

    private fun triggerUpvoteOnCollect() {
        // do an upvote if possible
        voteController.onVoteClicked?.invoke(Vote.UP)
    }

    /**
     * Displays the given [com.pr0gramm.app.feed.FeedItem] along with
     * the given vote.

     * @param item The item to display
     * *
     * @param vote The vote that belongs to the given item.
     */
    fun setFeedItem(item: FeedItem, isSelfPost: Boolean, vote: Vote) {
        val isSameItem = feedItem?.id == item.id

        this.feedItem = item
        this.isSelfPost = isSelfPost

        // update the views!
        usernameView.setUsername(item.user, item.mark)

        usernameView.setOnClickListener {
            onDetailClickedListener?.onUserClicked(item.user)
        }

        voteController.updateVote(vote, animate = isSameItem)

        collectionView.itemId = item.id

        updateViewState(vote)
    }

    private fun formatScore(vote: Vote): String? {
        val feedItem = this.feedItem ?: return null

        return if (scoreIsVisible()) {
            val rating = feedItem.up - feedItem.down + min(1, vote.voteValue)
            rating.toString()
        } else {
            null
        }
    }

    private fun scoreIsVisible(): Boolean {
        return isOneHourOld || isSelfPost || admin
    }

    /**
     * Updates the rating using the currently set feed item and the given vote.

     * @param vote The vote that is currently selected.
     */
    @SuppressLint("SetTextI18n")
    private fun updateViewState(vote: Vote) {
        val feedItem = this.feedItem ?: return

        // TODO hide the vote view (and follow state too?)
        // val viewVisibility = if (feedItem.deleted) View.INVISIBLE else View.VISIBLE
        // voteView.visibility = viewVisibility

        val textColor = captionView.currentTextColor

        val textSize = captionView.textSize.toInt()
        val offset = context.dp(2)

        val dClock = DrawableCache.get(R.drawable.ic_clock, textColor).withInsets(bottom = offset)
        dClock.setBounds(0, 0, textSize, textSize + offset)

        val dPlus = DrawableCache.get(R.drawable.ic_plus, textColor).withInsets(bottom = offset)
        dPlus.setBounds(0, 0, textSize, textSize + offset)

        ViewUpdater.replaceText(captionView, feedItem.created) {
            val date = DurationFormat.timeSincePastPointInTime(context, feedItem.created, short = true)
            val score = formatScore(vote)

            buildString {
                if (score != null) {
                    append("$score Benis · ")
                }

                append(date)
            }
        }
    }

    private fun followMenuClicked(selectedItemId: Int) {
        val state = when (selectedItemId) {
            R.id.action_follow_normal -> FollowState.FOLLOW
            R.id.action_follow_full -> FollowState.SUBSCRIBED
            else -> FollowState.NONE
        }
        followStateView.animate()
                .rotationYBy(360f)
                .setDuration(250L)
                .start()

        // update view
        updateFollowState(state)

        whileIsAttachedScope {
            withErrorDialog {
                // publish follow state to backend
                onDetailClickedListener?.updateFollowUser(state)
            }
        }

        // show a small hint that this is only viewable with pr0mium
        val userService: UserService = context.injector.instance()

        if (state != FollowState.NONE && !userService.canViewCategoryStalk) {
            Snackbar.make(this@InfoLineView, R.string.hint_follow_premium_only, Snackbar.LENGTH_SHORT)
                    .configureNewStyle()
                    .show()
        }
    }

    fun updateFollowState(followState: FollowState?) {
        if (followState == null) {
            followStateView.isVisible = false
            return
        }

        followStateView.isVisible = true

        when {
            followState.subscribed -> {
                val color = context.getColorCompat(ThemeHelper.accentColor)
                followStateView.setImageDrawable(DrawableCache.get(R.drawable.ic_action_follow_full, color))
            }

            followState.following -> {
                val color = context.getColorCompat(ThemeHelper.accentColor)
                followStateView.setImageDrawable(DrawableCache.get(R.drawable.ic_action_follow_normal, color))
            }

            else -> {
                val color = context.getStyledColor(android.R.attr.textColorSecondary)
                followStateView.setImageDrawable(DrawableCache.get(R.drawable.ic_action_follow_off, color))
            }
        }
    }

    private val isOneHourOld: Boolean
        get() {
            val oneHourAgo = Instant.now() - Duration.hours(1)
            return feedItem?.created?.isBefore(oneHourAgo) ?: false
        }
}
